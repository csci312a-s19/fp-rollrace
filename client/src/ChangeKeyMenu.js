import React from 'react';

const MENU_HEIGHT = 550;
const MENU_WIDTH = 700;

export default function Map(props) {
  // TODO: THIS LOOKS UGLY SO WE NEED TO RETHINK WITH ASPECT RATIO
  return (
    <g>
      <rect
        x={0}
        y={0}
        height={4000}
        width={4000}
        fill={'black'}
        opacity={0.5}
      />
      <rect
        rx={30}
        ry={30}
        x={(window.innerWidth - MENU_WIDTH / 2) / 2}
        y={(window.innerHeight - MENU_HEIGHT / 2) / 4}
        height={MENU_HEIGHT}
        width={MENU_WIDTH}
        fill={'white'}
        opacity={1}
      />
      <text
        x={(window.innerWidth - MENU_WIDTH / 2) / 2 + 300}
        y={(window.innerHeight - MENU_HEIGHT / 2) / 4 + 300}
      >
        Press New Jump Key
      </text>
      />
    </g>
  );
}
