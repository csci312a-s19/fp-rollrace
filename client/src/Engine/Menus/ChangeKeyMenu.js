/*
 * This file contains the ChangeKeyMenu component, which is called from the
 * paused menu. Key is not actually changed in GameEngine until the user exits.
 */

import React, { Component } from 'react';
import styled from 'styled-components';
import { ModalProvider } from 'styled-react-modal';
import PropTypes from 'prop-types';
import { StyledModal, ModalStyledButton } from '../../Style/ModalStyle.js';

const StyledP = styled.p`
  text-align: center;
  color: white;
  font-family: 'Gugi', cursive;
`;

class ChangeKeyMenu extends Component {
  constructor(props) {
    super(props);
    this._isMounted = false;

    /*
     * We have to maintain a separate state for jump key here and the GameEngine's
     * jump key becuase here we also display the current selected jump key on
     * top of the previous jump key.  The importance of this is that if the
     * player is to rapidly change key, there are many function calls, which
     * causes uneccesary computation.
     */
    this.state = {
      jumpKey: undefined
    };
    this.handleKeyChange = this.handleKeyChange.bind(this);
    this.handleExit = this.handleExit.bind(this);
  }

  componentDidMount() {
    this._isMounted = true;
    document.body.addEventListener('keypress', e => this.handleKeyChange(e));
  }

  // now we reset the jump key in GameEngine upon leaving
  handleExit() {
    const newKey = this.state.jumpKey;
    this.props.goBack(newKey);
  }

  handleKeyChange(event) {
    if (this._isMounted) {
      this.setState({ jumpKey: event.keyCode });
    }
  }

  componentWillUnmount() {
    document.body.removeEventListener('keypress', e => this.handleKeyChange(e));
    this._isMounted = false;
  }

  render() {
    const { showModal, currentKey } = this.props;
    const { jumpKey } = this.state;
    return (
      <ModalProvider className="modalprovider">
        <StyledModal className="styledmodal" isOpen={showModal}>
          <StyledP className="instruction">Press New jumpKey</StyledP>
          <StyledP className="current">
            {`Current Key: ${
              currentKey === 32
                ? 'SPACE'
                : String.fromCharCode(currentKey).toUpperCase()
            }`}
          </StyledP>
          {jumpKey && (
            <StyledP className="selected">
              {`Selected Key: ${
                jumpKey === 32
                  ? 'SPACE'
                  : String.fromCharCode(jumpKey).toUpperCase()
              }`}
            </StyledP>
          )}
          <ModalStyledButton className="back" onClick={this.handleExit}>
            Back
          </ModalStyledButton>
        </StyledModal>
      </ModalProvider>
    );
  }
}

ChangeKeyMenu.propTypes = {
  goBack: PropTypes.func.isRequired,
  showModal: PropTypes.bool.isRequired,
  currentKey: PropTypes.number.isRequired
};

export default ChangeKeyMenu;
