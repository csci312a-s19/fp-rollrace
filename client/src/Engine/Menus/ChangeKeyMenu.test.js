import React from 'react';
import ChangeKeyMenu from './ChangeKeyMenu.js';
import { shallow, mount } from 'enzyme';

describe('Change key menu tests', () => {
  const goBack = jest.fn();
  const showModal = true;
  const currentKey = 32;
  let comp;

  // Snapshot for reference
  describe('Change Key menu snapshot test', () => {
    test('progressBar snapshot', () => {
      comp = shallow(
        <ChangeKeyMenu
          goBack={goBack}
          showModal={showModal}
          currentKey={currentKey}
        />
      );
      expect(comp).toMatchSnapshot();
    });
  });

  describe('Change Key menu test functionality', () => {
    beforeEach(() => {
      comp = mount(
        <ChangeKeyMenu
          goBack={goBack}
          showModal={showModal}
          currentKey={currentKey}
        />
      );
    });

    // test props are same as what we assigned
    test('Prop testing', () => {
      expect(comp.props().showModal).toEqual(true);
      expect(comp.props().currentKey).toEqual(32);
    });

    // tests that all the elements are rendered on the component
    test('Model Components are in place', () => {
      expect(comp.find('.modalprovider').exists()).toEqual(true);
      expect(comp.find('.styledmodal').exists()).toEqual(true);
      expect(comp.find('.instruction').exists()).toEqual(true);
      expect(comp.find('.current').exists()).toEqual(true);
      expect(comp.find('.selected').exists()).toEqual(false);
    });

    // Test the back button on the component
    test('Component renders the back button', () => {
      expect(comp.find('.back').exists()).toEqual(true);
    });

    test('Clicking back button calls  correct callback function', () => {
      const backButton = comp
        .find('.back')
        .filterWhere(n => n.name() === 'button');
      backButton.simulate('click');
      expect(goBack).toHaveBeenCalledTimes(1);
    });
  });
});
