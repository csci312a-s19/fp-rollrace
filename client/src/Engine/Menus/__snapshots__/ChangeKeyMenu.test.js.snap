// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Change key menu tests Change Key menu snapshot test progressBar snapshot 1`] = `
<t
  className="modalprovider"
>
  <i
    className="styledmodal"
    isOpen={true}
  >
    <styled.p
      className="instruction"
    >
      Press New jumpKey
    </styled.p>
    <styled.p
      className="current"
    >
      Current Key: SPACE
    </styled.p>
    <styled.button
      className="back"
      onClick={[Function]}
    >
      Back
    </styled.button>
  </i>
</t>
`;
